lnd.go

```diff
@@ -630,105 +668,394 @@ func Main(cfg *Config, lisCfg ListenerCfg, implCfg *ImplementationCfg,
+
+	// Return a function closure that can be used to listen on a given
+	// address with the current TLS config.
+	restListen := func(addr net.Addr) (net.Listener, error) {
+		// For restListen we will call ListenOnAddress if TLS is
+		// disabled.
+		if t.cfg.DisableRestTLS {
+			return lncfg.ListenOnAddress(addr)
+		}
+
+		return lncfg.TLSListenOnAddress(addr, t.tlsCfg)
+	}
+
+	return serverOpts, restDialOpts, restListen, cleanUp, nil
+}
+
+// generateOrRenewCert generates a new TLS certificate if we're not using one
+// yet or renews it if it's outdated.
+func (t *TLSManager) generateOrRenewCert() error {
+	// Genereate a TLS pair if we don't have one yet.
+	err := t.createTLSPair()
+	if err != nil {
+		return err
+	}
+
+	certData, parsedCert, err := cert.LoadCertFromBytes(
+		t.certBytes, t.keyBytes,
+	)
+	if err != nil {
+		return err
+	}
+	t.certData = &certData
+	t.parsedCert = parsedCert
+
+	// Check to see if the certificate needs to be renewed.
+	err = t.certMaintenance()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// If a TLS pair doesn't exist yet, we create them and write them to disk. And
+// we encrypt the pair if needed. If the TLS pair already exists, we load it
+// into memory.
+func (t *TLSManager) createTLSPair() error {
+	// If TLS Key Encryption is on but the keyring is empty then we need to
+	// generate a temporary certificate.
+	var emptyKeyRing keychain.KeyRing
+	if t.cfg.TLSEncryptKey && (t.keyRing == emptyKeyRing) {
+		rpcsLog.Infof("Generating ephemeral TLS certificates...")
+
+		tmpValidity := validityHours * time.Hour
+		// Append .tmp to the end of the cert for differentiation.
+		tmpCertPath := t.cfg.TLSCertPath + ".tmp"
+		t.certPath = tmpCertPath
+		// Pass in a blank string for the key path so the
+		// function doesn't write them to disk.
+		certBytes, keyBytes, err := cert.GenCertPair(
+			"lnd temporary autogenerated cert", tmpCertPath,
+			"", t.cfg.TLSExtraIPs, t.cfg.TLSExtraDomains,
+			t.cfg.TLSDisableAutofill, tmpValidity,
+		)
+		if err != nil {
+			return err
+		}
+		t.certBytes = certBytes
+		t.keyBytes = keyBytes
+
+		err = cert.WriteCertPair(tmpCertPath, "", certBytes, keyBytes)
+		if err != nil {
+			return err
+		}
+
+		rpcsLog.Infof("Done generating ephemeral TLS certificates")
+	} else {
+		// Ensure we create TLS key and certificate if they don't exist.
+		if !fileExists(t.cfg.TLSCertPath) &&
+			!fileExists(t.cfg.TLSKeyPath) {
+
+			rpcsLog.Infof("Generating TLS certificates...")
+
+			certBytes, keyBytes, err := cert.GenCertPair(
+				"lnd autogenerated cert", t.cfg.TLSCertPath,
+				t.cfg.TLSKeyPath, t.cfg.TLSExtraIPs,
+				t.cfg.TLSExtraDomains, t.cfg.TLSDisableAutofill,
+				t.cfg.TLSCertDuration,
+			)
+			if err != nil {
+				return err
+			}
+			t.certBytes = certBytes
+			t.keyBytes = keyBytes
+
+			err = cert.WriteCertPair(
+				t.cfg.TLSCertPath, t.cfg.TLSKeyPath, certBytes,
+				keyBytes,
+			)
+			if err != nil {
+				return err
+			}
+
+			if t.cfg.TLSEncryptKey {
```

Comment:
As the linter also complains, this is a three level nested `if` and will be very hard to understand by anyone looking at it. Can you please refactor this part before we put any review work into this?

Reply:
Yes, makes sense, absolutely. Spent some time refactoring that.